nexus/src/app/instance.rs

```diff
@@ -1357,6 +1384,127 @@ impl super::Nexus {
         Ok(())
     }
 
+    /// Compares the existing Propolis record in CRDB (`db_instance`) with a new
+    /// instance runtime state supplied from a sled agent, determines whether
+    /// any Propolis VMMs will be left unused as a result of this transition,
+    /// and releases the resources that were reserved to them.
+    async fn reclaim_unused_propolis_resources(
+        &self,
+        opctx: &OpContext,
+        new_runtime: &nexus::InstanceRuntimeState,
+        db_instance: &nexus_db_model::Instance,
+    ) -> Result<(), Error> {
+        let log = &self.log;
+        let instance_id = db_instance.id();
+
+        // Handle changes to the instance's active Propolis ID. The only
+        // expected change is from the current ID to a migration destination ID.
+        let old_propolis_id = db_instance.runtime().propolis_id;
+        if new_runtime.propolis_id != old_propolis_id {
+            match db_instance.runtime().dst_propolis_id {
+                Some(dst_id) if dst_id != new_runtime.propolis_id => {
+                    warn!(log,
+                          "active Propolis did not change to migration target";
+                          "instance_id" => %instance_id,
+                          "new_propolis_id" => %new_runtime.propolis_id,
+                          "old_propolis_id" => %old_propolis_id,
+                          "dst_propolis_id" => %dst_id);
+                }
+                None => {
+                    warn!(log,
+                          "active Propolis changed without an active migration";
+                          "instance_id" => %instance_id,
+                          "new_propolis_id" => %new_runtime.propolis_id,
+                          "old_propolis_id" => %old_propolis_id);
+                }
+                Some(_) => {}
+            }
+
+            // Propolises stop themselves after a migration out, but there is no
+            // way to guarantee this has happened by the time this function
+            // executes (because this may have been triggered by an update from
+            // a migration target). To ensure the source Propolis is no longer
+            // consuming resources, explicitly unregister it before deleting
+            // its reservation.
+            info!(log, "reclaiming resources from migrated-out Propolis";
+                  "instance_id" => %instance_id,
+                  "propolis_id" => %old_propolis_id);
+
+            let old_sa =
+                self.sled_client(&db_instance.runtime().sled_id).await?;
+
+            // The caller is going to replace the instance runtime state with
+            // the newly-advanced state it supplied, so there's no need to write
+            // the returned runtime state back to CRDB here.
+            let _ = old_sa.instance_unregister(&instance_id).await?;
```

Comment:
Just to confirm, it's always the new sled-agent that would've triggered this nexus call right? i'd have to double check but if we got an update from the old sled-agent and tried to call back to it we might dead lock if it's holding the instance lock in sled-agent.

Reply:
Ooh, good catch.

Yes, right now it's always the new sled agent that triggers this call, though that's somewhat accidental--the update procedure is supposed to allow both sides to send the "transfer" state change, but the source doesn't do it because (today) the `InstanceRuntimeState` doesn't contain enough information about the destination for the source to be able to create the correct record. If it *did* have that information, though, we'd run right into the problem you're describing, because the monitor thread indeed holds the instance lock while publishing to Nexus.

In general, I'd prefer not to have the sled agent hold locks while calling out to other services. I think that's relatively doable in the monitor task's context. It might be more difficult in the `propolis_ensure` path, but that's less pressing for this PR because those calls to `publish_state_to_nexus` are explicitly only used when starting an instance without migrating it. I'll see if I can tighten up the monitor loop.